pkg/grpc/api.go

```diff
@@ -718,20 +674,27 @@ func (a *api) SaveState(ctx context.Context, in *runtimev1pb.SaveStateRequest) (
 		return empty, err
 	}
 
-	reqs := []state.SetRequest{}
-	for _, s := range in.States {
-		key, err1 := stateLoader.GetModifiedStateKey(s.Key, in.StoreName, a.id)
-		if err1 != nil {
-			return empty, err1
+	l := len(in.States)
+	if l == 0 {
+		return empty, nil
+	}
+
+	reqs := make([]state.SetRequest, l)
+	for i, s := range in.States {
+		var key string
+		key, err = stateLoader.GetModifiedStateKey(s.Key, in.StoreName, a.id)
+		if err != nil {
+			return empty, err
 		}
 		req := state.SetRequest{
 			Key:      key,
 			Metadata: s.Metadata,
 		}
 
-		if contentType, ok := req.Metadata[contribMetadata.ContentType]; ok && contentType == contenttype.JSONContentType {
-			if err1 = json.Unmarshal(s.Value, &req.Value); err1 != nil {
-				return empty, err1
+		if req.Metadata[contribMetadata.ContentType] == contenttype.JSONContentType {
```

Comment:
When getting from a map such as:

```go
v, ok := mymap[key]
```

If the key doesn't exist, `ok` is `false` and `v` is the zero value. The reason for the `ok` is to detect when the key is present BUT its value is the zero value.

In this case `ok && contentType == contenttype.JSONContentType` was a tautology. Whether the key was missing, or present but with a zero value, it would have failed `== contenttype.JSONContentType` regardless.